### ISSUE TITLE
Incorrect alignment for non-power-of-2 __vector types

### ISSUE THREAD

**Issue by tsbockman** [ldc-developers/ldc#3107]
writeln(__vector(float[3]).alignof);

The code above prints 12, but it should be 4 since alignments must be powers of 2. This likely breaks code both inside and outside of the compiler in surprising ways.

**Comment by kinke** [comment]
Thx. Seems to be due to a missing `TypeVector.alignment()` override in the front-end, a non-issue for DMD as it only supports 128 & 256 bit vector types.

**Comment by kinke** [comment]
> but it should be 4 since alignments must be powers of 2

Should it be 4 or 16? LLVM apparently uses a natural size of 16 for `float3` and size 32 for `float5`, so seems to round it up to the next power-of-2. If users want to store arrays of these odd vector types, they probably don't expect any padding.

**Comment by kinke** [comment]
By the looks of it, we can't make a `float3` only 12 bytes small; that wouldn't allow us to use a `<3 x float>` field in the IR struct type for `struct S { __vector(float[3]) v; }` anymore. So I guess size and alignment need to be rounded up to the next power of 2.

**Comment by baziotis** [comment]
I don't if it is related to this or maybe I miss something. When trying to compile:
```
void foo()
{
    alias V = __vector(int[3]);
    V v;
}
```
While in [godbolt compiles](https://godbolt.org/z/_reFiz), using my (aka github) version breaks:
```
ldc2: ../lib/IR/Instructions.cpp:1104: void llvm::AllocaInst::setAlignment(unsigned int): Assertion `(Align & (Align-1)) == 0 && "Alignment is not a power of 2!"' failed.
```

**Comment by kinke** [comment]
That's because LDC releases are built with an LLVM without assertions (for speed). The many useful assertions are the reason why we're using an LLVM *with* assertions for some CI systems (Azure & Shippable); we've caught quite a few LDC bugs this way.

**Comment by baziotis** [comment]
Ah, yes. But aren't assertions used for things that _must_ be true ? Unless they're also used as warnings in LLVM and so compilation can continue if some are false.

**Comment by JohanEngelen** [comment]
The assertions in LLVM are for things that _must_ be true, but the assertions are not meant for error reporting to the user. They are a debugging tool. Assertions are added liberally in LLVM, nice for bug hunting, not nice for execution performance.

**Comment by dnadlinger** [comment]
As in, compilation only continues in release mode because this happens not to cause a crash elsewhere, but the outcome is still undefined (and it's a bug in LDC to emit IR that violates the constraints set by LLVM).

**Comment by baziotis** [comment]
@JohanEngelen I understand what you said, but I think correctness nowadays becomes all the more important.
@dnadlinger your answer helped me. How it will continue with a false assertion is undefined. Which means one of:
- For me the best case: The compilation will continue and LLVM will crash.
- The compilation will continue and LLVM won't crash. One case here is that correct code will be generated. But the other and _worst_ case for me is that compilation will continue generating wrong code. Meaning, the generated program is will crash.

The last case I think is quite bad. On one hand, we can go with the "The programmer knows what they're doing" motto which exists in C, C++ and D so it makes some sense. But to me it seems weird to have a possible warning be generated for wrong code and trade it with compile-time performance.

**Comment by dnadlinger** [comment]
Of course it is bad, but that's because LLVM expects the compiler frontends not to generate invalid IR in the first place. In assertions-enabled mode, more of these preconditions are verified at run-time to aid in debugging, but at quite a hefty performance cost.

This doesn't really have anything to do with "The programmer knows what they are doing" – there is a number of other possible logic bugs in a compiler, which can cause, to LLVM, perfectly valid IR to be generated, which still does not reflect the semantics of the source language program the programmer wrote.

**Comment by baziotis** [comment]
> This doesn't really have anything to do with "The programmer knows what they are doing" – there is a number of other possible logic bugs in a compiler, which can cause, to LLVM, perfectly valid IR to be generated, which still does not reflect the semantics of the source language program the programmer wrote.

I agree and the motto irrelevant is irrelevant to this indeed. I mean that this motto stands because for example using a `void3` vector is a programmer error. That we could have prevented. Although, you might mean that "yes we could have prevent it and we should prevent it, but this is a job of the front-end, not the assertions of the back-end".

**Comment by dnadlinger** [comment]
> Although, you might mean that "yes we could have prevent it and we should prevent it, but this is a job of the front-end, not the assertions of the back-end".

Yes, it's a job of the frontend indeed to emit an error message with proper location information, etc. – especially because D also supports things like `__traits(compile, …)`, and thus a backend ICE is always a compiler bug.

**Comment by baziotis** [comment]
Ok, clear. I agree!
